botwo/botwo.py

```diff
@@ -0,0 +1,200 @@
+        self.config = None
+
+    def build(self, mapping, config):
+        if not isinstance(mapping, dict):
+            raise TypeError("Invalid client mapping type: " + str(type(mapping)) + " expected dict")
+
+        if not mapping.get("default"):
+            raise ValueError("default client is required")
+        self.mapping = mapping
+        self.config = config
+        self.default = mapping.get("default")
+
+        for k, v in self.mapping.items():
+            if not isinstance(v, botocore.client.BaseClient):
+                raise TypeError("mapping: " + k + "Invalid client type: " + str(type(v)) + " expected boto.s3.client")
+
+        for profile in self.config:
+            if not self.mapping.get(profile):
+                raise ValueError("profile: " + profile + " mapping is required")
+
+        class_attributes = self._create_methods()
+        cls = type("s3", (), class_attributes)
+        return cls()
+
+    def _create_methods(self):
+        op_dict = {}
+        operations = [func for func in dir(self.default) if
+                      (callable(getattr(self.default, func)) and not func.startswith('_'))]
+        for operation_name in operations:
+            if operation_name == "get_paginator":
+                op_dict[operation_name] = self._create_get_paginate_method(operation_name)
+            elif operation_name == "can_paginate":
+                op_dict[operation_name] = self._create_can_paginate_method(operation_name)
+            elif operation_name == "delete_objects":
+                op_dict[operation_name] = self._create_delete_objects_method(operation_name)
+            elif operation_name in LIST_METHODS:
+                op_dict[operation_name] = self._create_list_method(operation_name)
+            elif operation_name in COPY_METHODS:
+                op_dict[operation_name] = self._create_copy_method(operation_name)
+            else:
+                op_dict[operation_name] = self._create_api_method(operation_name)
+        op_dict["meta"] = self.default.meta
+        return op_dict
+
+    def _create_api_method(self, operation_name):
+        def _api_call(_, *args, **kwargs):
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            client_to_call = self.default
+            if kwargs.get("Bucket"):  # bucket operation
+                res = self._route_params(kwargs)
+                client_to_call = res[0]
+                kwargs = res[1]
+
+            return getattr(client_to_call, operation_name)(**kwargs)
+
+        _api_call.__name__ = str(operation_name)
+        return _api_call
+
+    def _create_list_method(self, operation_name):
+        def _api_call(_, *args, **kwargs):
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            client_to_call = self.default
+            if kwargs.get("Bucket") and not kwargs.get("Prefix"):
```

Comment:
Do we truly want to ignore empty `Prefix` (as opposed to calls where `Prefix` is not specified)? AFAIU this will do the wrong thing if I try to list with an empty `Prefix`.

This pattern of calling `kwargs.get(k)` and ignoring the return value instead of `k in kwargs` also appears in other places, and I am not sure it is correct.  (But I am not a Python expert, and will be happy to learn that I am incorrect in thinking that these are errors!)